### ISSUE TITLE
`az container exec` has no way to pass arguments

### ISSUE THREAD

**Issue by elibarzilay** [Azure/azure-cli#10039]
There is no way to specify arguments in `az container exec`.  For example:

    az container exec -g ... -n ... --exec-command "ls /"

will just complain about not finding the `ls /` executable.  There should really be some way to pass arguments instead of using a shell for doing anything.

(As it is right now, the comment "The most common use case is to open an interactive bash shell" is really misleading, since running a shell is just about the only useful thing you can do when you cannot pass arguments.)

(This bug is a superset of #8687)

**Comment by yonzhan** [comment]
container

**Comment by navba-MSFT** [comment]
@elibarzilay Apologies for the late reply. Could you refer [this](https://docs.microsoft.com/en-us/cli/azure/container?view=azure-cli-latest#az-container-exec) article and let me know if using the [most recent version](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli) of the CLI helps is resolving this issue ? Awaiting your reply.

**Comment by navba-MSFT** [comment]
@elibarzilay I wanted to do quick follow-up to check if you had a chance to look at my above comment. This github issue is open for quite sometime. Please let us know if you need any further assistance on this. Awaiting your reply.

**Comment by speak2beeb** [comment]
@navba-MSFT , Yes, I can confirm that this is still a bug with the latest az version.  Notice:

`$ az version
{
  "azure-cli": "2.33.1",
  "azure-cli-core": "2.33.1",
  "azure-cli-telemetry": "1.0.6",
  "extensions": {
    "aks-preview": "0.5.48"
  }
}

$ az container exec -n XX -g XX --container-name XX --exec-command "ls"
file  file  file

$ az container exec -n XX -g XX --container-name XX --exec-command "ls -l"
OCI runtime exec failed: exec failed: container_linux.go:380: starting container process caused: exec: "ls -l": executable file not found in $PATH: unknown`

Any idea when we could have a fix to this?

**Comment by navba-MSFT** [comment]
@speak2beeb Since the image details are not mentioned here, it would be difficult to provide a plan. Nonetheless, I am sending the below command if you could try that and then run the` ls -l` command from the _cli shell_ it should help.

>az container exec -g data-repo --name mycontainer1 **--exec-command "/bin/bash"**

More Info [here](https://docs.microsoft.com/en-us/azure/container-instances/container-instances-exec#run-a-command-with-azure-cli).

**Comment by speak2beeb** [comment]
Apologies. The image I am using is Microsoft's, namely: mcr.microsoft.com/azure-pipelines/vsts-agent

As for the "/bin/bash" command, this gives me an interactive session. The requirement is for the command to be executed non-interactively. (Clearly, the "ls -l" was just an example to show that any command with a space does not work.)

I hope this brings more clarity to this defect.

**Comment by speak2beeb** [comment]
Can we get an update on this?

**Comment by navba-MSFT** [comment]
@speak2beeb Thanks for getting back. Apologies for the delay. I was able to run the exec command on my container with "ls -l". Please find below the steps I followed:

>az container create --resource-group MyContainerRG --name myapp1 --image mcr.microsoft.com/azure-pipelines/vsts-agent --command-line "tail -f /dev/null"

**Note**: In the above command, I have specified the "tail -f /dev/null" to keep the connection alive. Else the container will terminate. More Info [here](https://docs.microsoft.com/en-us/azure/container-instances/container-instances-troubleshooting#container-continually-exits-and-restarts-no-long-running-process).

Then I ran the below exec command, in both _windows box_ as well as _cloud shell_ and it ran successfully.
>az container exec -g MyContainerRG --name myapp1 --exec-command "ls -l" 

**Output:** 
>total 22872
-rw-r--r-- 1 root root 23415665 Jan 25  2016 phantomjs-2.1.1-linux-x86_64.tar.bz2
-rwxrwxr-x 1 root root     2659 Jan 15  2019 start.sh

Hope this helps.

**Comment by speak2beeb** [comment]
Same error, sir:
$ az container exec -n XX -g XX --container-name XX --exec-command "tail -f /dev/null"
**OCI runtime exec failed: exec failed: container_linux.go:380: starting container process caused: exec: "tail -f /dev/null": stat tail -f /dev/null: no such file or directory: unknown**

Also adding this note: I upgraded my local azure-cli from version 2.33.1 to 2.34.1 and still have the same error.

**Comment by navba-MSFT** [comment]
@speak2beeb Could you please check my above command again ? I created the container first using **--command-line** and then ran the **exec** command.

**Comment by speak2beeb** [comment]
Hello Msft,
I've ran the following commands and still getting the same error.

```
$ az container create -g XX -n testsboxaci --image mcr.microsoft.com/azure-pipelines/vsts-agent --cpu 1 --memory 4 --location EastUS --command-line "tail -f /dev/null" --environment-variables VSTS_ACCOUNT=YY VSTS_TOKEN=ZZ VSTS_AGENT=testsboxaci VSTS_POOL=ACI_POOL
$ az container exec -n testsboxaci -g XX --exec-command "ls -l"
  OCI runtime exec failed: exec failed: container_linux.go:380: starting container process caused: exec: "ls -l": executable file not found in $PATH: unknown
```

You mentioned above "I have specified the 'tail -f /dev/null' to keep the connection alive." This doesn't quite make sense since the container is running fine. That is why I'm able to successfully run _any_ command without a space (such as "ls").

**Comment by jwshive** [comment]
I'm running into this issue as well today trying to run a sql script on a msft sql server container. 

`=> az container exec -g <resource-group> -n <container-name> --exec-command "/opt/mssql-tools/bin/sqlcmd -S localhost -U SA -P <password> -i /opt/scripts/create_database.sql"
OCI runtime exec failed: exec failed: container_linux.go:380: starting container process caused: exec: "/opt/mssql-tools/bin/sqlcmd -S localhost -U SA -P <password> -i /opt/scripts/create_database.sql": stat /opt/mssql-tools/bin/sqlcmd -S localhost -U SA -P <password> -i /opt/scripts/creater_database.sql: no such file or directory: unknown`

**Comment by jwshive** [comment]
Keep the issue open.

**Comment by navba-MSFT** [comment]
@speak2beeb @jwshive @elibarzilay I am tagging the Service Team to look into this issue.

@macolso Could you please look into this issue and provide an update once you get a chance ? Awaiting your reply.

**Comment by speak2beeb** [comment]
Could we get an update on this? This would be immensely useful for our enterprise if we can get this defect fixed.

**Comment by bantoni** [comment]
I tried to execute commands using az ci exec on a newly created Azure Container Instance running mcr.microsoft.com/azuredocs/aci-helloworld:latest
It works if I've assigned a public IP address. **If I use a private IP it fails** with the mentioned error:
az container exec -g anb --name anbci --container-name anbci --exec-command "ls -la"        
OCI runtime exec failed: exec failed: container_linux.go:380: starting container process caused: exec: "ls -l": executable file not found in $PATH: unknown

**Comment by teriansilva** [comment]
The issue still occurs as of today. How is the status on it?

**Comment by bantoni** [comment]
Tried it again and now it works for me.

**Comment by speak2beeb** [comment]
Yes! I can confirm this now works! Clearly they've fixed it somewhere along the way.
This ticket can be closed! :)

**Comment by tblazejowski** [comment]
I may confirm this issue **still occurs** while trying to execute below command inside pipeline's **AzureCLI@2** task:
`az container exec --exec-command "<any command with argument>" --resource-group <resource_group> --name <container_group_name>`

**Comment by Capriolo88** [comment]
I can confirm that the issue is still present.
I executed the command on a running container in West Europe, but the cli still see the --exec-command value as a single command without considering it as command and arguments

az container exec --container-name <container_name> --name <name> --resource-group <rg> --exec-command "ls -la"                                                                                           OCI runtime exec failed: exec failed: container_linux.go:380: starting container process caused: exec: "ls -la": stat ls -la: no such file or directory: unknown

**Comment by inakianduaga** [comment]
Run into this just now. Solution for me was to move the command from the `command` parameter to rather specify it via a container `environment variable`. I created a helper script (well ChatGpt did)

```sh
  #!/bin/bash
# This command needs to be included inside the Dockerfile so we can point to it via Azure container instance commmand, which can't take parameters
# https://github.com/Azure/azure-cli/issues/10039
# https://stackoverflow.com/questions/68095848/azure-container-instance-command-execution-returns-no-such-file-or-directory

set -euo pipefail

# Get the command from the environment variable
COMMAND="$DOCKER_CONTAINER_RUN_COMMAND"

# Check if the command is provided
if [ -z "$COMMAND" ]; then
  echo "Error: Environment variable DOCKER_CONTAINER_RUN_COMMAND is not set."
  exit 1
fi

# Execute the command using /bin/sh -c
/bin/sh -c "$COMMAND"
```

that I put into a file `run-command-from-environment.sh`

which gets added to the dockerfile upon container image building

```dockerfile
...
COPY ./run-command-from-environment.sh <SOME_PATH_IN_YOUR_CONTAINER_IMAGE_FILESYSTEM> 
```

And then the azure container instance is configured to (I'm using terraform CDK but you get the point)

```ts
    new ContainerGroup(this, `container-run-pipeline-${pipelineName}`, {
      // ...
      container: [
        {
          //...
          environmentVariables: {
            DOCKER_CONTAINER_RUN_COMMAND: 'some-command-with-whatever-parameters-you-want-to-run'
          },
          commands: ["./run-command-from-environment.sh"],
        },
      ],
    });
```

hope it helps someone
